src/ast/nodes/ConditionalExpression.ts

```diff
@@ -23,62 +23,73 @@ export default class ConditionalExpression extends NodeBase {
 		callback: ForEachReturnExpressionCallback,
 		options: ExecutionPathOptions
 	) {
-		if (this.hasUnknownTestValue) {
+		const testValue = this.hasUnknownTestValue ? UNKNOWN_VALUE : this.getTestValue();
+		if (testValue === UNKNOWN_VALUE || testValue) {
 			this.consequent.forEachReturnExpressionWhenCalledAtPath(path, callOptions, callback, options);
+		}
+		if (testValue === UNKNOWN_VALUE || !testValue) {
 			this.alternate.forEachReturnExpressionWhenCalledAtPath(path, callOptions, callback, options);
-		} else {
-			this.forEachRelevantBranch(node =>
-				node.forEachReturnExpressionWhenCalledAtPath(path, callOptions, callback, options)
-			);
 		}
 	}
 
 	getValue(): any {
-		const testValue = this.test.getValue();
+		const testValue = this.hasUnknownTestValue ? UNKNOWN_VALUE : this.getTestValue();
 		if (testValue === UNKNOWN_VALUE) return UNKNOWN_VALUE;
 		return testValue ? this.consequent.getValue() : this.alternate.getValue();
 	}
 
 	hasEffects(options: ExecutionPathOptions): boolean {
-		return (
-			this.test.hasEffects(options) ||
-			(this.hasUnknownTestValue
-				? this.consequent.hasEffects(options) || this.alternate.hasEffects(options)
-				: this.someRelevantBranch(node => node.hasEffects(options)))
-		);
+		if (this.test.hasEffects(options)) return true;
+		const testValue = this.hasUnknownTestValue ? UNKNOWN_VALUE : this.getTestValue();
+		if (testValue === UNKNOWN_VALUE) {
+			return this.consequent.hasEffects(options) || this.alternate.hasEffects(options);
+		}
+		return testValue ? this.consequent.hasEffects(options) : this.alternate.hasEffects(options);
 	}
 
 	hasEffectsWhenAccessedAtPath(path: ObjectPath, options: ExecutionPathOptions): boolean {
```

Comment:
I've always wondered what this means - is it referring to `obj.path[c ? d : e]`? Or something else?

Reply:
No, it means you access e.g. `(c ? d : e).the.path`. In this example, the path `['the', 'path']` of `c ? d : e` is accessed.

For `obj.path[c ? d : e]`, we would just be accessing the path `[UNKNOWN_KEY]` of `obj.path`. The "unknown" is because non-literal computed keys are never evaluated at the moment (we might extend `getValueAtPath` to be used here as well, though).